BUG: CatboostClassifier explanations with feature interactions fail on Windows

### Issue Description

This a pretty strange bug that I encountered [here](https://github.com/shap/shap/actions/runs/6458284627/job/17531689364?pr=3318#step:6:116) that only appears under Windows.


### Minimal Reproducible Example

```python
import catboost

X, y = shap.datasets.adult(n_points=50)
model = catboost.CatBoostClassifier(depth=1, iterations=10).fit(X, y)
predicted = model.predict(X, prediction_type="RawFormulaVal")

ex_cat = shap.TreeExplainer(model)

# catboost explanations
explanation = ex_cat(X, interactions=True)
```


### Traceback

```shell
def _validate_inputs(self, X, y, tree_limit, check_additivity):
        # see if we have a default tree_limit in place.
        if tree_limit is None:
            tree_limit = -1 if self.model.tree_limit is None else self.model.tree_limit

        if tree_limit < 0 or tree_limit > self.model.values.shape[0]:
>           tree_limit = self.model.values.shape[0]
E           AttributeError: 'TreeEnsemble' object has no attribute 'values'
```


### Expected Behavior

Should result in the same as under linux.

The problem here is with the tempfile creation

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

shap master

--- a/shap/explainers/_tree.py
+++ b/shap/explainers/_tree.py
@@ -6,6 +6,7 @@ import warnings
 import numpy as np
 import pandas as pd
 import scipy.special
+import os
 from packaging import version
 
 from .. import maskers
@@ -1837,10 +1838,11 @@ class CatBoostTreeModelLoader:
         # cb_model.save_model("cb_model.json", format="json")
         # self.loaded_cb_model = json.load(open("cb_model.json", "r"))
         import tempfile
-        tmp_file = tempfile.NamedTemporaryFile()
+        tmp_file = tempfile.NamedTemporaryFile(delete=False)
         cb_model.save_model(tmp_file.name, format="json")
         self.loaded_cb_model = json.load(open(tmp_file.name))
         tmp_file.close()
+        os.unlink(tmp_file.name)
 
         # load the CatBoost oblivious trees specific parameters
         self.num_trees = len(self.loaded_cb_model['oblivious_trees'])

--- a/shap/explainers/_tree.py
+++ b/shap/explainers/_tree.py
@@ -1,4 +1,5 @@
 import json
+import os
 import struct
 import time
 import warnings
@@ -6,7 +7,6 @@ import warnings
 import numpy as np
 import pandas as pd
 import scipy.special
-import os
 from packaging import version
 
 from .. import maskers

--- a/shap/explainers/_tree.py
+++ b/shap/explainers/_tree.py
@@ -1835,14 +1835,11 @@ class XGBTreeModelLoader:
 
 class CatBoostTreeModelLoader:
     def __init__(self, cb_model):
-        # cb_model.save_model("cb_model.json", format="json")
-        # self.loaded_cb_model = json.load(open("cb_model.json", "r"))
         import tempfile
-        tmp_file = tempfile.NamedTemporaryFile(delete=False)
-        cb_model.save_model(tmp_file.name, format="json")
-        self.loaded_cb_model = json.load(open(tmp_file.name))
-        tmp_file.close()
-        os.unlink(tmp_file.name)
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            tmp_file = os.path.join(tmp_dir, "model.json")
+            cb_model.save_model(tmp_file, format="json")
+            self.loaded_cb_model = json.load(open(tmp_file))
 
         # load the CatBoost oblivious trees specific parameters
         self.num_trees = len(self.loaded_cb_model['oblivious_trees'])

--- a/tests/explainers/test_tree.py
+++ b/tests/explainers/test_tree.py
@@ -387,7 +387,7 @@ def test_catboost_categorical():
 
 
 def test_catboost_interactions():
-    # GH 3187
+    # GH 3324
     catboost = pytest.importorskip("catboost")
 
     X, y = shap.datasets.adult(n_points=50)

--- a/tests/explainers/test_tree.py
+++ b/tests/explainers/test_tree.py
@@ -386,6 +386,24 @@ def test_catboost_categorical():
     )
 
 
+def test_catboost_interactions():
+    # GH 3187
+    catboost = pytest.importorskip("catboost")
+
+    X, y = shap.datasets.adult(n_points=50)
+
+    model = catboost.CatBoostClassifier(depth=1, iterations=10).fit(X, y)
+    predicted = model.predict(X, prediction_type="RawFormulaVal")
+
+    ex_cat = shap.TreeExplainer(model)
+
+    # catboost explanations
+    explanation = ex_cat(X, interactions=True)
+    assert (
+        np.abs(explanation.values.sum(axis=(1, 2)) + explanation.base_values - predicted).max()
+        < 1e-4
+    )
+
 # TODO: Test tree_limit argument
 
 

